lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+                                        TupleTree<model::Binary> &Binary) const;
+};
+
+std::optional<model::TypePath>
+Converter::toCABI(const model::RawFunctionType &Function,
+                  TupleTree<model::Binary> &Binary) const {
+  revng_log(Log, "Converting a `RawFunctionType` to `CABIFunctionType`.");
+  revng_log(Log, "Function:\n" << serializeToString(Function));
+
+  if (!ABI.checkCompatibility(Function.Arguments(), Function.ReturnValues())) {
+    revng_log(Log,
+              "FAIL: the function is not compatible with `"
+                << model::ABI::getName(ABI.ABI()) << "`.");
+    return std::nullopt;
+  }
+
+  // A single place to conveniently store all the types that need to be added
+  // to the model in order for the new function prototype to be valid.
+  auto M = manageTypeDependencies(*Binary);
+
+  // Convert register arguments first.
+  auto RegArguments = M.tryConvertingRegisterArguments(Function.Arguments());
+  if (RegArguments == std::nullopt) {
+    revng_log(Log, "FAIL: Register argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // Then stack arguments.
+  auto Stack = M.tryConvertingStackArguments(Function.StackArgumentsType(),
+                                             RegArguments->size());
+  if (Stack == std::nullopt) {
+    revng_log(Log, "FAIL: Stack argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // And the return type.
+  auto ReturnType = M.tryConvertingReturnValue(Function.ReturnValues());
+  if (ReturnType == std::nullopt) {
+    revng_log(Log, "FAIL: Return value conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // If all the partial conversions passed, the conversion is guaranteed to be
+  // successful. All the `TypeDependencies` can now be recorded into the
+  // model.
+  M.commit();
+
+  // With all the dependencies out of the way, new `CABIFunctionType` can now
+  // be created,
+  auto [NewType, NewTypePath] = Binary->makeType<model::CABIFunctionType>();
+  NewType.ID() = Function.ID();
+  NewType.CustomName() = Function.CustomName();
+  NewType.OriginalName() = Function.OriginalName();
+  NewType.ABI() = ABI.ABI();
+
+  // And filled in with the argument information.
+  using model::Argument;
+  for (auto &Argument : llvm::concat<Argument>(*RegArguments, *Stack))
+    NewType.Arguments().insert(std::move(Argument));
+  NewType.ReturnType() = *ReturnType;
+
+  revng_log(Log, "Conversion successful.");
+
+  // To finish up the conversion, remove all the references to the old type by
+  // carefully replacing them with references to the new one.
+  return replaceAllUsesWith(Function.key(), NewTypePath, Binary);
+}
+
+using TDManager = TypeDependencyManager;
+template<typename RT>
+std::optional<llvm::SmallVector<model::Argument, 8>>
+TDManager::tryConvertingRegisterArguments(const SortedVector<RT> &Used) {
```

Comment:
Why do you call them `Used`?

Reply:
Because in one of the previous designs I didn't have enough space to give the argument a proper name 😅, and it's kind of left this way.

For context, this function used to take two different register lists: the ones function actually "used" and the ones that were "allowed" by the ABI. The second one got assimilated into the runtime "no-longer-trait" and is implicitly available, so there's no reason to pass them anymore.

It can now be called just `Registers` or something like that.

Reply:
Done. See the `(not reviewed)` commits.